### PR TITLE
docs: Upgrading k8s components

### PR DESCRIPTION
**Pull request by aaronlevy**
/cc @robszumski @joshix @bcwaldon

### CODE REVIEW

File: Documentation/kubernetes-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+* kubelet
+* kube-proxy
+* kube-apiserver
+* kube-controller-manager
+* kube-scheduler
+* kube-podmaster (high-availability)
+
+### Upgrading kube-apiserver and kube-proxy
+
+Both the kube-apiserver and kube-proxy are run as "static pods". This means the pod definition is a file on disk (default location: `/etc/kubernetes/manifests`). To update these components, you simply need to update the static manifest file.
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: Perhaps say something here about relying on the kubelet to respawn pods in response to manifest changes.

File: Documentation/kubernetes-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+For example, to upgrade the kube-apiserver version, you could make the following change to: `/etc/kubernetes/manifests/kube-apiserver.yaml`
+
+```
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: This formatting is weird. Just put the `image: ` lines in pre blocks

File: Documentation/kubernetes-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+The kube-controller-manager, kube-scheduler, and kube-podmaster are all also deployed as static pods in `/etc/kubernetes/manifests`. However, in high-availability deployments, the kube-podmaster is responsible for making sure only a single copy of the controller-manager and scheduler are running cluster-wide.
+
+To accomplish this the kube-podmaster on each node, if master-elected, will copy the static manifest from `/srv/kubernetes/manifets` into `/etc/kubernetes/manifests` and the kubelet will pick up the manifest and run the pod. If the kube-podmaster loses its status as master, it will remove the static pod from `/etc/kubernetes/manifests/` and the kubelet will shut down the pod.
+
+This configuration means upgrading of these components will take a little more coordination.
+
+To upgrade the kube-controller-manager and kube-scheduler:
+
+1. Make changes to the base manifests in `/srv/kubernetes/manifests`
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: This is only true for the current master - how does this differ for inactive masters?

File: Documentation/kubernetes-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well.
+
+If you wish to run a custom version of the kubelet binary, the kubelet service file (`/etc/systemd/system/kubelet.service`) should be modified to use the custom binary path on each node (master & worker).
+
+## Upgrading Master Nodes
+
+Master nodes will consist of the following Kubernetes components:
+
+* kubelet
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: You should just drop this from the list given the disclaimer above about how the kubelet is managed.

File: Documentation/kubernetes-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Upgrading Kubernetes
+
+This document will describe upgrading your cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.0/docs/admin/cluster-management.html).
+
+## Upgrading the Kubelet
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well.
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: Maybe expand this to say "this step will not be covered below in the master- or worker-specific documentation"

File: Documentation/kubernetes-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+1. Remove the existing manifests (if present) from `/etc/kubernetes/manifests`
+1. The kube-podmaster will automatically fetch the new manifest from `/srv/kubernetes/manifests` and copy it to `/etc/kubernetes/manifests` and the new pod will be started.
+
+### Upgrading Worker Nodes
+
+Worker nodes will consist of the following kubernetes components.
+
+* kubelet
+* kube-proxy
+
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: What does a user actually have to do? Anything? kube-proxy was never covered

File: Documentation/kubernetes-upgrade.md (hunk 1 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: ack

File: Documentation/kubernetes-upgrade.md (hunk 2 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: ack

File: Documentation/kubernetes-upgrade.md (hunk 3 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: The manifests should be changed in all masters, active or not (so if master changes, it's still  using updated manifest). But I can make a note about this.

File: Documentation/kubernetes-upgrade.md (hunk 6 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: Woops, got a little aggressive with deletes at the bottom. I'll add it back.

File: Documentation/kubernetes-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+## Upgrading the Kubelet
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well.
+
+If you wish to run a custom version of the kubelet binary, the kubelet service file (`/etc/systemd/system/kubelet.service`) should be modified to use the custom binary path on each node (master & worker).
+
+## Upgrading Master Nodes
+
+Master nodes will consist of the following Kubernetes components:
```
Comment (robszumski) [inline Documentation/kubernetes-upgrade.md]: Might be good to mention that user pods will continue to run on the workers as the masters are updated.

File: Documentation/kubernetes-upgrade.md (hunk 3 above)
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: Right - the inactive masters won't copy the updated manifests from /srv to /etc, for example. I wanted to make that point clear.

File: Documentation/kubernetes-upgrade.md (hunk 4 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: ack

File: Documentation/kubernetes-upgrade.md (hunk 7 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: Gonna get more specifics on eviction/timeouts but I'll add something here.

Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: Added note below

File: Documentation/kubernetes-upgrade.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
+# Upgrading Kubernetes
+
+This document will describe upgrading your cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.0/docs/admin/cluster-management.html).
```
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: This document describes upgrading the Kubernetes components on a cluster's master and worker nodes.

File: Documentation/kubernetes-upgrade.md (hunk 7 above)
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: Master nodes consist of ...

File: Documentation/kubernetes-upgrade.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+## Upgrading the Kubelet
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well. This step will not be covered in the guides below, however, more information can be found in the [CoreOS Updates Documentation](https://coreos.com/using-coreos/updates)
+
+If you wish to run a custom version of the kubelet binary, the kubelet service file (`/etc/systemd/system/kubelet.service`) should be modified to use the custom binary path on each node (master & worker).
```
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: To run a custom version of the kubelet, modify the kubelet service file on each node (`/etc/systemd/system/kubelet.service`) to contain the path to the custom kubelet binary.

File: Documentation/kubernetes-upgrade.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
+# Upgrading Kubernetes
+
+This document will describe upgrading your cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.0/docs/admin/cluster-management.html).
+
+## Upgrading the Kubelet
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well. This step will not be covered in the guides below, however, more information can be found in the [CoreOS Updates Documentation](https://coreos.com/using-coreos/updates)
```
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: The Kubelet binary is part of CoreOS, and is automatically upgraded along with the operating system. Upgrading it is not covered here. More information can be found in the [CoreOS Updates Documentation](https://coreos.com/using-coreos/updates).

File: Documentation/kubernetes-upgrade.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+This document describes upgrading the Kubernetes components on a cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.0/docs/admin/cluster-management.html).
+
+**NOTE:** The following upgrade documentation is for installations based on the CoreOS + Kubernetes step-by-step [installation guide](https://coreos.com/kubernetes/docs/latest/getting-started.html). Upgrade documentation for the AWS cloud-formation based installation is forthcoming.
+
+## Upgrading the Kubelet
+
+The Kubelet binary ships with CoreOS releases. As the host OS is updated, the Kubelet will be upgraded as well. This step will not be covered in the guides below, however, more information can be found in the [CoreOS Updates Documentation](https://coreos.com/using-coreos/updates)
+
+To run a custom version of the kubelet, modify the kubelet service file on each node (`/etc/systemd/system/kubelet.service`) to contain the path to the custom kubelet binary..
```
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: Typo: two periods..

File: Documentation/kubernetes-upgrade.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# Upgrading Kubernetes
+
+This document describes upgrading the Kubernetes components on a cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.0/docs/admin/cluster-management.html) and [version upgrade notes](http://kubernetes.io/v1.0/docs/design/versioning.html#upgrades)
```
Comment (joshix) [inline Documentation/kubernetes-upgrade.md]: Noting that this doc specifically points to the v1.0 k8s docs.
Is this desirable or should we amend now to v1.1?
The http://k8s.io/v/docs/ paths don't have a URL scheme for making links to "/latest/" (that I've deciphered yet). That is inconvenient.

Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: Prob makes sense to point to 1.1 docs

**Comment by joshix** [comment]
Related https://github.com/coreos/bugs/issues/978

File: Documentation/kubernetes-upgrade.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
+# Upgrading Kubernetes
+
+This document describes upgrading the Kubernetes components on a cluster's master and worker nodes. For general information on Kubernetes cluster management, upgrades (including more advanced topics such as major API version upgrades) see the [Kubernetes upstream documentation](http://kubernetes.io/v1.1/docs/admin/cluster-management.html) and [version upgrade notes](http://kubernetes.io/v1.1/docs/design/versioning.html#upgrades)
+
+**NOTE:** The following upgrade documentation is for installations based on the CoreOS + Kubernetes step-by-step [installation guide](https://coreos.com/kubernetes/docs/latest/getting-started.html). Upgrade documentation for the AWS cloud-formation based installation is forthcoming.
+
+## Upgrading the Kubelet
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: I think it's worth noting in this section that it applies to both master and worker nodes. I suspect that some folks don't think of a master as being a "kubelet".

File: Documentation/kubernetes-upgrade.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+The kube-proxy is run as a "static pod". To upgrade the pod definition, simply modify the pod manifest located in `/etc/kubernetes/manifests/kube-proxy.yaml`. The kubelet will pick up the changes and re-launch the kube-proxy pod.
+
+## Example Upgrade Process:
+
+1. Prepare new pod manifests for master nodes
+1. Prepare new pod manifests for worker nodes
+1. For each master node:
+    1. Back up existing manifests
+    1. Shut down the kubelet service
```
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: Why do you need to stop/start the kubelet?

Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: You don't have to, the kubelet will see the changed file and reload the static manifests. This is mostly me being overly cautious, but I don't feel strongly about this so I can remove or leave it.

File: Documentation/kubernetes-upgrade.md (hunk 13 above)
Comment (aaronlevy) [inline Documentation/kubernetes-upgrade.md]: added below

File: Documentation/kubernetes-upgrade.md (hunk 14 above)
Comment (bcwaldon) [inline Documentation/kubernetes-upgrade.md]: The kubelet is designed to watch for changes on the filesystem, and we direct folks to rely on that above, so I don't see a need to kick the service.

**Comment by bcwaldon** [comment]
LGTM after language regarding kubelet restart is removed.

**Pull request merged**